From 8fc64d7997b191364e722a46e53523e79a5f43c4 Mon Sep 17 00:00:00 2001
From: chandavath <rcrajagopal@habana.ai>
Date: Tue, 20 Feb 2024 06:43:11 +0200
Subject: [PATCH] [SW-137122] gaudi3: erom rec: added hard reset support

added hard reset support in erom recovery. Clear the
recovery flag and disable pcie and perform hard reset.

Change-Id: Ic3254d1dd1dea0b62e080e57f2f0d180fef3bfde
Signed-off-by: chandavath <rcrajagopal@habana.ai>
---

diff --git a/arch/gaudi3/common/reset.c b/arch/gaudi3/common/reset.c
index ec9fc6f..80d0d80 100644
--- a/arch/gaudi3/common/reset.c
+++ b/arch/gaudi3/common/reset.c
@@ -2,7 +2,41 @@
 /*
  * Copyright (C) 2023 HabanaLabs Ltd.
  */
-/*
- * TODO: Add gaudi3 specific eROM implementation
- * Added empty source for gaudi3 eROM compilation .
- */
+
+#include <baremetal.h>
+#include <gaudi3_blocks_linux_driver.h>
+#include <gaudi3_reg_map.h>
+#include <global_conf_regs.h>
+#include <global_conf_masks.h>
+#include <psoc_reset_conf_regs.h>
+#include <pll.h>
+#include <uart.h>
+
+#define DELAY_1K_USEC	1000
+
+static void hl_rst_sw_all_reset(void)
+{
+	preboot_puts("Issuing hard reset\n");
+
+	regmap_rmwritel(D0_PSOC_GLOBAL_CONF_REG(mmGLOBAL_CONF_DIE_READY_IN), 0x0,
+					GLOBAL_CONF_DIE_READY_IN_IND_M);
+	/* Issue reset to slave die */
+	regmap_writel(0x1, D1_PSOC_RESET_CONF_REG(mmPSOC_RESET_CONF_SW_ALL_RST));
+	/* wait for slave die to reset no indication is required */
+	udelay(DELAY_1K_USEC);
+	/* Issue master die0 reset */
+	regmap_writel(0x1, D0_PSOC_RESET_CONF_REG(mmPSOC_RESET_CONF_SW_ALL_RST));
+}
+
+void hl_dev_reset(void)
+{
+	preboot_recovery_flag(false);
+
+	hl_pcie_disable_access();
+
+	/* TODO: DISABLE SEC RELATED ACCESS */
+
+	hl_rst_sw_all_reset();
+
+	wfe();
+}
